core/src/main/scala/com/rallyhealth/vapors/core/dsl/factfilter/Dsl.scala

```diff
@@ -40,34 +40,35 @@ private[dsl] class Dsl extends Types with Syntax {
 
   // TODO: Figure out how to use this conditional tertiary expression
 
-  def when[T, A](exp: AnyExp[T, Boolean])(thenExp: AnyExp[T, A])(elseExp: AnyExp[T, A]): AnyExp[T, A] = liftAnyExp {
+  def when[T, A](exp: Exp[T, Boolean])(thenExp: Exp[T, A])(elseExp: Exp[T, A]): Exp[T, A] = liftExp {
     ExpAlg.Cond[T, A](exp, thenExp, elseExp)
   }
 
-  def withFactsOfType[U : ClassTag : TypeTag](factType: FactType[U]): TypedFactExpBuilder[Any, U] = {
-    withFactsOfTypeIn[Any, U](FactTypeSet.of(factType))
+  def withFactsOfType[T >: U, U : ClassTag : TypeTag](factType: FactType[U]): WhereFactsExpBuilder[T, U] = {
+    new WhereFactsExpBuilder[T, U](FactTypeSet.of(factType))
   }
 
-  def withFactsOfTypeIn[T >: U, U : ClassTag : TypeTag](factTypeSet: FactTypeSet[U]): TypedFactExpBuilder[T, U] = {
-    new TypedFactExpBuilder[T, U](factTypeSet)
+  // TODO: Use shapeless for this
```

Comment:
Remove?  Or planning to do this?

Reply:
I think I went with removing type parameters to avoid needing shapeless. Restricting input types makes the interface harder to use without much value. My gamble is that any expression at this level will require some form of type selection because relying on the Scala type is not specific enough to select the fact types you would want to filter to. For example, you might have "weight in lbs" represented in some Scala type, but that doesn't tell you where that measurement came from. You might want a fact representing "self-reported weight" and another representing "weight measurement at the doctor's office". Both are the same type of value, but they are different types of facts. Shapeless can't really help us here, so the payoff for using it throughout this level of the expression doesn't really buy us much.

There may be other use cases where shapeless would be a good fit in the future, but I don't think we need it yet.